src/oci/mod.rs

```diff
@@ -0,0 +1,143 @@
+use std::fmt;
+use std::time::Duration;
+
+use bollard::Docker;
+
+pub use bollard::auth::DockerCredentials as Credentials;
+pub use bollard::errors::Error as ConnectionError;
+
+mod image;
+pub use image::{Image, LocalImage};
+
+mod models;
+pub use models::{ImageUri, InvalidImageUriError};
+
+mod registry;
+pub use registry::{RegistryAuth, RegistryAuthClient, RegistryAuthError};
+
+#[derive(Debug, Clone)]
+pub struct Client(Docker);
+
+impl Client {
+    /// Connect to the daemon based on the `DOCKER_HOST` environment variable.
+    pub async fn connect() -> Result<Self> {
+        let inner = Docker::connect_with_defaults()?;
+
+        // Bollard doesn't actually connect with the `connect_*` call.
+        // Do a /ping to ensure we can connect before proceeding.
+        inner
+            .ping()
+            .await
+            .map_err(Error::with_context("failed to connect to daemon"))?;
+
+        // TODO: determine which engine it is
+        // let version_info = inner
+        //     .version()
+        //     .await
+        //     .map_err(Error::with_context("failed to fetch version info"))?;
+
+        Ok(Self(inner))
+    }
+
+    #[inline]
+    fn inner(&self) -> &Docker {
+        &self.0
+    }
+
+    #[inline]
+    fn inner_mut(&mut self) -> &mut Docker {
+        &mut self.0
+    }
+
+    /// Sets the request timeout, affecting all subsequent requests to the daemon.
+    /// Default is 2 minutes.
+    pub fn set_timeout(&mut self, timeout: Duration) {
+        self.inner_mut().set_timeout(timeout);
+    }
+
+    /// Sets the request timeout, affecting all subsequent requests to the daemon.
+    /// Default is 2 minutes.
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.set_timeout(timeout);
+        self
+    }
+
+    /// Exposes methods to work with images.
+    #[inline]
+    pub fn image(&self) -> Image<'_> {
+        Image::new(self)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(c) = &self.1 {
+            c.fmt(f)?;
+            ": ".fmt(f)?;
+        }
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub struct Error(ConnectionError, Option<String>);
```

Comment:
[thiserror](https://github.com/dtolnay/thiserror) will implement the [source](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) method of Error if the property name is `source`. Perhaps we could define this as follows, which also makes it a bit more readable?

 ```suggestion
pub struct Error {
    context: Option<String>
    source: ConnectionError
}
```